### DISCUSSION TITLE
UART PAL requirements in multi-threaded OS

### DISCUSSION THREAD

**Discussion by jkcko** [openthread/openthread#9873] [Q&A] [answered]
In a multi-threaded OS environment where where UART TX and RX processing may be done in a separate thread is it okay to call otPlatUartSendDone() and otPlatUartReceived() from another thread than the main OpenThread thread which calls otTaskletsProcess?

**Comment by abtink** [comment, marked as answer]
Generally no.  All public OT APIs and platform callbacks should be called from same OS context (task/thread).

You can refer to the specific API documentation to see if it explicitly allows calls from different contexts. I know of only one example `otPlatSpiSlaveTransactionCompleteCallback`:
https://github.com/openthread/openthread/blob/f0b6fcea6ef77c9a54ab11767593f9a8798e3662/include/openthread/platform/spi-slave.h#L67-L70)
